### ISSUE TITLE
Segmentation fault occurs when sequence array element is undefined or partially defined

### ISSUE THREAD

**Issue by Orbuild** [eclipse-cyclonedds/cyclonedds#1994]
- When I use the code from the master branch, I ran a simple publish-subscribe program, where the message format is as follows:
```
module MyTestData
{ 
  struct Msg {
    @key long id;
    sequence<octet> data[10];
  };
};
```
- If the sequence array is not defined or only some elements in the array are defined, the segmentation fault will occur in the program. I would like to ask if this is the agreed behavior (all elements in the sequence array need to be defined), or is this a problem? Looking forward to your response!
```
for (int i = 0; i < 5; i++) {
    msg->data[i]._buffer = dds_alloc(100);
    msg->data[i]._length = 100;
    msg->data[i]._release = true;
}
```
```
=== [MyTestPublisher]  Waiting for a reader to be discovered ...
[MyTestPublisher]  Writing : Segmentation fault (core dumped)
```

**Comment by eboasson** [comment]
Yes, in general samples that you pass into the API need to be valid. That means:

* strings must point to a null-terminated string or be a null pointer
* sequences need `_length` ≤ `_maximum` and `_buffer` may not be a null pointer if `_length` > 0
* the sequence buffer will not be allocated/resized/freed if `_release` is `false` *except* that if everything is 0 it will take over ownership in read/take and allocate and that it is somewhat forgiving of `_maximum` < `_length`
* all elements of an array need to be initialized
* the union case identified by the discriminant must be correctly initialized
* `@external` pointers may never be a null pointer on write.

Maybe I forgot something, maybe not 😂 I'm sure you understand. That exception noted above about sequences exists so you can memset everything 0 and you'll be good for read and take.

There are few functions that only care about the values of the key fields, for example `dds_dispose`. Those ignore all other fields, so in your example, you can call `dds_dispose` with `id` set correctly and everything else garbage.
